pkg/apis/pravega/v1beta1/pravegacluster_types.go

```diff
@@ -1194,6 +1206,66 @@ func (p *PravegaCluster) validateConfigMap() error {
 	return nil
 }
 
+func (p *PravegaCluster) ValidateSegmentStore() error {
+	totalMemoryLimitsQuantity := p.Spec.Pravega.SegmentStoreResources.Limits[corev1.ResourceMemory]
+	totalMemoryRequestsQuantity := p.Spec.Pravega.SegmentStoreResources.Requests[corev1.ResourceMemory]
+	if (resource.Quantity{}) == totalMemoryLimitsQuantity {
+		return fmt.Errorf("Missing required value for field spec.pravega.segmentStoreResources.limits.memory")
+	}
+
+	totalMemoryLimits := totalMemoryLimitsQuantity.Value()
+	totalMemoryRequests := totalMemoryRequestsQuantity.Value()
+	if totalMemoryLimits < totalMemoryRequests {
+		return fmt.Errorf("spec.pravega.segmentStoreResources.requests.memory value must be less than or equal to spec.pravega.segmentStoreResources.limits.memory")
+	}
+
+	cacheSizeString := p.Spec.Pravega.Options["pravegaservice.cache.size.max"]
+	if cacheSizeString == "" {
+		return fmt.Errorf("Missing required value for option pravegaservice.cache.size.max")
+	}
+	cacheSizeQuantity := resource.MustParse(cacheSizeString)
+	maxDirectMemoryString := ""
+	xmxString := ""
+
+	for _, value := range p.Spec.Pravega.SegmentStoreJVMOptions {
+		if strings.Contains(value, "-Xmx") {
+			xmxString = strings.ToUpper(strings.TrimPrefix(value, "-Xmx")) + "i"
+		}
+
+		if strings.Contains(value, "-XX:MaxDirectMemorySize=") {
+			maxDirectMemoryString = strings.ToUpper(strings.TrimPrefix(value, "-XX:MaxDirectMemorySize=")) + "i"
+		}
+	}
+
+	if xmxString == "" {
+		return fmt.Errorf("Missing required value for Segment Store JVM Option -Xmx")
+	}
+	xmxQuantity := resource.MustParse(xmxString)
+
+	if maxDirectMemoryString == "" {
+		return fmt.Errorf("Missing required value for Segment Store JVM option -XX:MaxDirectMemorySize")
+	}
+	maxDirectMemoryQuantity := resource.MustParse(maxDirectMemoryString)
+
+	xmx := xmxQuantity.Value()
+	maxDirectMemorySize := maxDirectMemoryQuantity.Value()
+	cacheSize := cacheSizeQuantity.Value()
+
+	if totalMemoryLimits <= maxDirectMemorySize+xmx {
+		return fmt.Errorf("MaxDirectMemorySize(%v B) along with JVM Xmx value(%v B) is greater than or equal to the total available memory(%v B)!", maxDirectMemorySize, xmx, totalMemoryLimits)
+	}
+
+	if maxDirectMemorySize <= xmx {
```

Comment:
Please, remove this part of the validation (also from the PR description), as this may be completely legitimate. You may want to configure a JVM of 4GB with a cache of 3GB (and therefore setting `maxDirectMemorySize=4GB`). And there is nothing strictly wrong with that, is just a user choice. We have to validate to prevent what we know for sure is wrong, but not more than that, as we also want to give some degree of flexibility to users.

Reply:
According to this [document](https://github.com/pravega/pravega/blob/master/documentation/src/docs/admin-guide/segmentstore-memory.md), we should generally provision the Segment Store with more direct memory than heap memory. So, shouldn't this validation be there. I agree with your example and can remove the `equality` from the check so as to provide flexibility to the user.